File: esw-ocs/esw-ocs-dsl-kt/src/main/kotlin/esw/ocs/dsl/internal/ScriptWiring.kt
/*
 * Copyright (c) 2022 Thirty Meter Telescope International Observatory
 * SPDX-License-Identifier: Apache-2.0
 */

package esw.ocs.dsl.internal

import esw.ocs.dsl.lowlevel.CswServices
import esw.ocs.dsl.script.StrandEc
import esw.ocs.impl.script.ScriptContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.channels.Channel
import java.time.Duration

class ScriptWiring(val scriptContext: ScriptContext) {
    val heartbeatInterval: Duration by lazy { scriptContext.heartbeatInterval() }
    val strandEc: StrandEc by lazy { StrandEc.apply() }
    private val supervisorJob by lazy { SupervisorJob() }
    private val dispatcher by lazy { strandEc.executorService().asCoroutineDispatcher() }
    val scope: CoroutineScope by lazy { CoroutineScope(supervisorJob + dispatcher) }
    val cswServices: CswServices by lazy { CswServices.create(scriptContext, strandEc) }
    val heartbeatChannel: Channel<Unit> by lazy { Channel() }

    fun shutdown() {
        supervisorJob.cancel()
        dispatcher.close()
        heartbeatChannel.cancel()
    }
}

File: esw-ocs/esw-ocs-dsl-scala/src/main/scala/esw/ocs/dsl2/internal/ScriptWiring.scala
package esw.ocs.dsl2.internal

import akka.actor.typed.ActorSystem
import csw.params.events.SequencerObserveEvent
import esw.ocs.dsl2.lowlevel.CswServices
import esw.ocs.dsl.script.{ScriptDsl, StrandEc}
import esw.ocs.dsl2.highlevel.{
  AlarmServiceDsl,
  CommandServiceDsl,
  ConfigServiceDsl,
  CswHighLevelDsl,
  DatabaseServiceDsl,
  EventServiceDsl,
  LocationServiceDsl,
  LoopDsl
}
import esw.ocs.impl.script.ScriptContext

import java.time.Duration
import java.util.concurrent.ScheduledExecutorService
import scala.concurrent.ExecutionContext
import scala.jdk.DurationConverters.JavaDurationOps

class ScriptWiring(scriptContext: ScriptContext) {
  val heartbeatInterval: Duration          = scriptContext.heartbeatInterval
  val strandEc: StrandEc                   = StrandEc()
  given ExecutionContext                   = strandEc.ec
  val dispatcher: ScheduledExecutorService = strandEc.executorService
  val cswServices: CswServices             = CswServices(scriptContext, strandEc)
  val heartbeatChannel                     = ???
  given ActorSystem[_]                     = scriptContext.actorSystem

  val scriptDsl: ScriptDsl = new ScriptDsl(
    scriptContext.sequenceOperatorFactory,
    scriptContext.jLogger,
    strandEc,
    () => shutdown()
  )

  val loopDsl = LoopDsl(strandEc)

  val cswHighLevelDsl: CswHighLevelDsl = CswHighLevelDsl(
    scriptContext.jLogger.asScala,
    cswServices,
    LocationServiceDsl(cswServices.locationService),
    ConfigServiceDsl(cswServices.configService),
    EventServiceDsl(cswServices.eventPublisher, cswServices.eventSubscriber),
    CommandServiceDsl(),
    AlarmServiceDsl(
      cswServices.alarmService,
      scriptContext.config.getDuration("csw-alarm.refresh-interval").toScala,
      loopDsl
    ),
    cswServices.timeService,
    DatabaseServiceDsl(cswServices.databaseServiceFactory, cswServices.locationService),
    loopDsl,
    scriptContext,
    strandEc,
    SequencerObserveEvent(scriptContext.prefix)
  )

  def shutdown(): Unit = {
    strandEc.shutdown()
  }
}
